### PR TITLE
HWY-171: Include list of votes' hashes that need endorsing in `WireVote`.

### PR DESCRIPTION
**Pull request by goral09**
https://casperlabs.atlassian.net/browse/HWY-171

### CODE REVIEW

File: node/src/components/consensus/highway_core/highway/vertex.rs [hunk 1]
```diff
@@ -148,6 +149,7 @@ impl<C: Context> Debug for WireVote<C> {
             .field("timestamp", &self.timestamp.millis())
             .field("panorama", self.panorama.as_ref())
             .field("round_exp", &self.round_exp)
+            .field("endorsed_votes", &self.endorsed)
```
Comment (afck) [inline node/src/components/consensus/highway_core/highway/vertex.rs:152]: Let's keep the `Debug` impl in line with the code as much as possible:
```suggestion
            .field("endorsed", &self.endorsed)
```

File: node/src/components/consensus/highway_core/state.rs [hunk 2]
```diff
@@ -254,8 +259,12 @@ impl<C: Context> State<C> {
     }
 
     /// Returns hash of vote that needs to be endorsed.
-    pub(crate) fn needs_endorsements(&self, _vote: &SignedWireVote<C>) -> Option<C::Hash> {
-        None
+    pub(crate) fn needs_endorsements(&self, vote: &SignedWireVote<C>) -> Option<C::Hash> {
+        vote.wire_vote
+            .endorsed
+            .iter()
+            .find(|hash| self.endorsements.contains_key(&hash))
```
Comment (afck) [inline node/src/components/consensus/highway_core/state.rs:266]: Did you mean
```suggestion
            .find(|hash| !self.endorsements.contains_key(&hash))
```
?
This should return the _missing_ dependency, should it?

File: node/src/components/consensus/highway_core/active_validator.rs [hunk 3]
```diff
@@ -295,6 +295,7 @@ impl<C: Context> ActiveValidator<C> {
             panorama = state.panorama().clone();
         }
         let seq_number = panorama.next_seq_num(state, self.vidx);
+        let endorsed = state.endorsements().collect();
```
Comment (afck) [inline node/src/components/consensus/highway_core/active_validator.rs:298]: Maybe add a TODO that this is a placeholder: Once the LNC is implemented, only some of our known endorsements will be dependencies of the new vote.

File: node/src/components/consensus/highway_core/test_macros.rs [hunk 4]
```diff
@@ -86,6 +87,30 @@ macro_rules! add_vote {
             seq_number,
             timestamp: ($time).into(),
             round_exp: $round_exp,
+            endorsed: vec![],
+        };
+        let hash = wvote.hash();
+        let swvote = SignedWireVote::new(wvote, &TestSecret(($creator).0), &mut $rng);
+        $state.add_vote(swvote).map(|()| hash)
```
Comment (afck) [inline node/src/components/consensus/highway_core/test_macros.rs:94]: Can this whole case be reduced to the next case (by just adding a semicolon?), and deduplicated?
```rust
add_vote!($state, $rng, $creator, $time, $round_exp, $val; $($obs),*;)
```

Comment (goral09) [reply node/src/components/consensus/highway_core/test_macros.rs:94]: I didn't want to replace all the 28 places where the `add_vote!`, without endorsements, is being called. I will see if your suggestion achieves that.

Comment (afck) [reply node/src/components/consensus/highway_core/test_macros.rs:94]: No, that's not what I mean!
I just mean replacing the _implementation_ of this case with that single line I wrote.

Comment (afck) [reply node/src/components/consensus/highway_core/test_macros.rs:94]: A macro can call itself.

Comment (goral09) [reply node/src/components/consensus/highway_core/test_macros.rs:94]: OK, it almost worked. I had to pass `vec![]` after the semicolon.

File: node/src/components/consensus/highway_core/active_validator.rs (hunk 3 above)
Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs:298]: I was actually thinking that `state.endorsements` would hold only votes that have to be endorsed according to the `state.panorama` so we would always take all there is and only update the `state.endorsements` when needed.

Comment (afck) [reply node/src/components/consensus/highway_core/active_validator.rs:299]: I hadn't thought about it that way… :thinking: 
Not sure that works, though: Endorsement requirements can basically go away again, once we cite a vote that cites the corresponding endorsements… or at least that makes them redundant and a bit wasteful.

**Review by afck — APPROVED** [review]

File: node/src/components/consensus/highway_core/active_validator.rs (hunk 3 above)
Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs:299]: > once we cite a vote that cites the corresponding endorsements

In this case I would remove those endorsements from `state.endorsements` after adding that vote to this node's protocol state. I will think more about it though.

**Pull request merged**